test/e2e/hooks_test.go

```diff
@@ -0,0 +1,105 @@
+//go:build functional
+// +build functional
+
+package e2e
+
+import (
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/suite"
+	v12 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
+	"github.com/argoproj/argo-workflows/v3/test/e2e/fixtures"
+)
+
+type HooksSuite struct {
+	fixtures.E2ESuite
+}
+
+func (s *HooksSuite) TestWorkflowLevelHooks() {
+	s.Given().
+		Workflow(`apiVersion: argoproj.io/v1alpha1
+kind: Workflow
+metadata:
+  generateName: lifecycle-hook-
+spec:
+  entrypoint: main
+  hooks:
+    exit:
+      template: http
+    running:
+      expression: workflow.status == "Running"
```

Comment:
interesting - this fires every time? I'm actualy not sure how to interpret this hook.

Reply:
it will fire only one time the first time of expression satisfy.